File: 剑指offer/56.1.数组中数字出现的次数.py
from typing import List


class Solution:
    def singleNumbers(self, nums: List[int]) -> List[int]:
        ret = 0

        for num in nums:
            ret = ret ^ num

        div = 1
        while div & ret == 0:
            div <<= 1

        a = 0
        b = 0

        for num in nums:
            if num & div:
                a = a ^ num
            else:
                b = b ^ num

        return [a, b]

File: 剑指offer/56.1.数组中数字出现的次数.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> singleNumbers(vector<int>& nums)
    {
        int ret = 0;
        for (int num : nums) {
            ret ^= num;
        }

        int index = 1;
        while (ret & index == 0) {
            index <<= 1;
        }

        int a = 0;
        int b = 0;

        for (int num : nums) {
            if (num & index) {
                a ^= num;
            } else {
                b ^= num;
            }
        }
        return { a, b };
    }
};